android/src/main/java/com/reactnativecommunity/webview/extension/file/Base64FileDownloader.kt

```diff
@@ -0,0 +1,169 @@
+			return
+		}
+		val (mimeType, extension) = getMimeTypeAndFileExtension(base64Parts[0])
+
+		showAlertDialog(context, extension) {
+			Toast.makeText(context, downloadingMessage, Toast.LENGTH_LONG).show()
+			val fileBytes = Base64.decode(base64Parts[1], Base64.DEFAULT)
+			saveFileToDownloadsFolder(context, base64, fileBytes, mimeType, extension, requestFilePermission)
+		}
+	}
+
+	/**
+	 * This method is called from PermissionListener after WRITE_EXTERNAL_STORAGE permission is granted
+	 */
+	fun downloadBase64FileWithoutPermissionCheckAndDialog(
+		context: Context,
+		base64: String,
+		downloadingMessage: String,
+	) {
+		val base64Parts = base64.split(",")
+		if (base64Parts.size != 2) {
+			Log.e("Base64FileDownloader", "Unable to parse base64 to download a file. Base64 value was: $base64")
+			return
+		}
+		val (mimeType, extension) = getMimeTypeAndFileExtension(base64Parts[0])
+		Toast.makeText(context, downloadingMessage, Toast.LENGTH_LONG).show()
+		val fileBytes = Base64.decode(base64Parts[1], Base64.DEFAULT)
+		saveFileToDownloadsFolder(context, base64, fileBytes, mimeType, extension, {})
+	}
+
+	private fun showAlertDialog(context: Context, extension: String, onPositiveButtonClick: () -> Unit) {
+		AlertDialog.Builder(context)
+			.apply {
+				setMessage("Do you want to download \nFile.${extension}?")
+				setCancelable(false)
+				setPositiveButton("Download", object : DialogInterface.OnClickListener {
+					override fun onClick(p0: DialogInterface?, p1: Int) {
+						onPositiveButtonClick()
+					}
+				})
+				setNegativeButton("Cancel", object : DialogInterface.OnClickListener {
+					override fun onClick(p0: DialogInterface?, p1: Int) {
+						// Do nothing
+					}
+				})
+			}
+			.create()
+			.show()
+	}
+
+
+	private fun getMimeTypeAndFileExtension(header: String): Pair<String, String> {
+		val mimeType = Regex("data:([^;]+)").find(header)?.groupValues?.get(1) ?: "application/octet-stream"
+		val extension = MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType) ?: "bin"
+		return mimeType to extension
+	}
+
+	private fun saveFileToDownloadsFolder(
+		context: Context,
+		base64: String,
+		fileBytes: ByteArray,
+		mimeType: String,
+		extension: String,
+		requestFilePermission: (String) -> Unit,
+	) {
+		try {
+			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+				saveFileForAndroidQAndLater(context, fileBytes, mimeType, extension)
+			} else {
+				saveFileForAndroidLowerQ(context, base64, fileBytes, extension, requestFilePermission)
+			}
+		} catch (e: Exception) {
+			Log.e("Base64FileDownloader", "Unable to download base64 file. Base64 value was: $base64", e)
+			Toast.makeText(context, "Invalid filename or type", Toast.LENGTH_LONG).show()
+		}
+	}
+
+	@RequiresApi(Build.VERSION_CODES.Q)
+	private fun saveFileForAndroidQAndLater(context: Context, fileBytes: ByteArray, mimeType: String, extension: String) {
+		val resolver = context.contentResolver
+		val fileName = "File.$extension"
```

Comment:
@Tyschenko If you have to use a placeholder filename then use `download.$extension`